File: scripts/extract_syllabus.py
# -*- coding: utf-8 -*-
"""OCTA -- read the CPE 412 teaching plan out of the syllabus DOCX.

WHY DOCX AND NOT THE PDF.

The PDF is a landscape table. `pdftotext -layout` interleaves the columns, so a
unit outcome comes out fused to an unrelated topic heading -- 93 fragments, none
cleanly transcribable, and chapter 18 vanished entirely because its topic cell is
empty. The DOCX keeps table cells as discrete XML. No reconstruction, no
guessing, no possibility of inventing content -- which matters, because hard rule
5 forbids exactly that.

THREE THINGS WORD DOES THAT NAIVE EXTRACTION GETS WRONG.

1. `cell.text` joins paragraphs and loses the bullet structure. Iterate
   `cell.paragraphs` instead.

2. `paragraph.text` SILENTLY DROPS <w:br/>. The syllabus uses soft breaks
   between chapter blocks, so whole topic groups ran together invisibly. Walk
   the XML and map <w:br/>, <w:cr/> and <w:tab/> to newlines.

3. The author typed several topics into ONE paragraph -- "Computer Memory System
   Overview" and "Characteristics of Memory Systems" share a line. Word kept
   them as separate RUNS, and the run boundary sits exactly on the semantic
   split. `cell_lines()` below recovers it, conservatively: see the four
   conditions there, each of which exists to prevent a false split that was
   observed in the real file.

Usage:  python scripts/extract_syllabus.py <path-to-docx>   -> JSON on stdout
"""

import io
import json
import re
import sys

import docx
from docx.oxml.ns import qn

# Windows defaults stdout to cp1252, which turns Intel's U+2019 apostrophe into
# U+FFFD before the caller ever sees it -- and that lands in a stage file.
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", newline="")

VERBS = (
    "Define", "Differentiate", "Illustrate", "Discuss", "Elaborate", "Outline",
    "Examine", "Identify", "Present", "Compute", "List", "Draw", "Compare",
    "Enumerate", "Explain", "Understand", "Summarize", "Describe", "Provide",
    "Distinguish", "Name", "State", "Analyze", "Apply", "Determine", "Classify",
    "Interpret", "Demonstrate", "Design", "Recall", "Solve", "Use", "Evaluate",
    "Assess", "Justify",
)

BREAKS = {qn("w:br"), qn("w:cr"), qn("w:tab")}


def norm(s):
    """Collapse runs of whitespace (NBSP included) and strip bullet glyphs."""
    s = re.sub(r"[^\S\n]+", " ", s.replace(" ", " "))
    return s.strip().strip("•●·-*– \t")


def run_text(run):
    """A run's text WITH its breaks. run.text drops <w:br/> silently."""
    buf = []
    for node in run._element.iter():
        if node.tag == qn("w:t"):
            buf.append(node.text or "")
        elif node.tag in BREAKS:
            buf.append("\n")
    return "".join(buf)


def cell_lines(cell, run_rule=False, protect_first=False):
    """Every bullet in a table cell, one per list entry.

    An explicit <w:br/> ALWAYS ends an entry. That alone is enough for every
    column except the topics outline, where the author typed several topics onto
    one line; `run_rule` turns on the extra recovery described below.

    THE RUN RULE IS FOR TOPIC LISTS ONLY. Outcomes are sentences, and running it
    there truncated "Discuss the evolution of Intel's Cache memory and ARM Cache
    Organization" at "...ARM Cache" -- the tail was emitted as its own entry
    and then dropped by the verb filter, silently losing an outcome. Title Case
    is a property of headings, not of prose, so the rule has no business in a
    prose column.

    `protect_first` additionally exempts the cell's FIRST paragraph. In the
    topics column that paragraph is the chapter title, and titles are the other
    place the rule reliably misfires: "Top Level View of Computer Function and |
    Interconnection" satisfies every condition and is still wrong.

    A run boundary splits only when ALL of these hold. Each condition exists
    because dropping it produced a wrong split in the real file:

      a. the next run starts with a capital letter  -- topics are Title Case
      b. the previous run ends with a literal space -- the author's separator
      c. the previous run is not whitespace-only    -- else "The | Memory
                                                      Hierarchy" splits, since
                                                      Word emits ' ' as its own
                                                      run constantly
      d. at least two words accumulated             -- a lone capitalised word
                                                      is never a topic

    Verified against chapter 4, whose topics exercise every branch:
    "Cache Size | Mapping Function" splits, "The Memory Hierarchy" does not.
    """
    out = []

    def flush(text):
        t = norm(text)
        if t:
            out.append(t)

    for pi, p in enumerate(cell.paragraphs):
        runs = p.runs
        cur = ""
        may_split = run_rule and not (protect_first and pi == 0)
        for i, run in enumerate(runs):
            t = run_text(run)

            if may_split and i > 0 and cur.strip():
                prev = run_text(runs[i - 1])
                if (
                    t[:1].isupper()                   # a
                    and prev.endswith(" ")            # b
                    and prev.strip() != ""            # c
                    and len(cur.split()) >= 2         # d
                ):
                    flush(cur)
                    cur = ""

            # An explicit break always ends the current topic.
            if "\n" in t:
                chunks = t.split("\n")
                cur += chunks[0]
                for c in chunks[1:-1]:
                    flush(cur)
                    cur = c
                if len(chunks) > 1:
                    flush(cur)
                    cur = chunks[-1]
            else:
                cur += t
        flush(cur)
    return out


def main():
    doc = docx.Document(sys.argv[1])
    rows = []
    for table in doc.tables:
        if len(table.columns) != 7:
            continue
        for r in table.rows:
            cells = r.cells
            outcomes = cell_lines(cells[1])
            topics = cell_lines(cells[2], run_rule=True, protect_first=True)
            if not topics:
                continue
            rows.append({
                "cilo": " ".join(cell_lines(cells[0])),
                # Unit outcomes are written verb-first; anything else in that
                # column is a heading or a stray, and is not an outcome.
                "outcomes": [o for o in outcomes if o.split()[0] in VERBS],
                "outcomes_raw": outcomes,
                "topics": topics,
                "time": " ".join(cell_lines(cells[6])),
            })
    print(json.dumps(rows, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: scripts/extract-syllabus.mjs
#!/usr/bin/env node
// OCTA -- extract the CPE 412 teaching plan from the syllabus DOCX.
//
// WHY DOCX AND NOT THE PDF.
//
// The PDF is a landscape table. `pdftotext -layout` interleaves the columns, so
// a unit outcome comes out fused to an unrelated topic heading:
//
//     "Differentiate Computer Organization 1.2 Structure and Function"
//
// 93 fragments, none cleanly transcribable. Rendering the pages to images and
// reading them visually would work, but it is slower and a vision model can
// hallucinate table structure -- a documented weakness for exactly this task.
//
// The DOCX keeps table cells as discrete XML. Column 1 is the unit outcomes,
// column 2 is the topics, and each bullet is its own paragraph. No ambiguity,
// no reconstruction, no possibility of inventing content -- which matters,
// because hard rule 5 forbids exactly that.
//
//   node scripts/extract-syllabus.mjs            print the plan
//   node scripts/extract-syllabus.mjs --json     machine-readable
//
// Requires: python with `python-docx` (already installed for this repo).

import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const DOCX = resolve(ROOT, "docs", "source", "CPE 412.docx");
const SCRIPT = resolve(ROOT, "scripts", "extract_syllabus.py");


if (!existsSync(DOCX)) {
  console.error(`\nSyllabus DOCX not found at ${DOCX}\n`);
  process.exit(1);
}

const raw = execFileSync("python", [SCRIPT, DOCX], {
  encoding: "utf8",
  maxBuffer: 32 * 1024 * 1024,
});
const rows = JSON.parse(raw);

if (process.argv.includes("--json")) {
  process.stdout.write(JSON.stringify(rows, null, 2));
  process.exit(0);
}

const c = {
  bold: (s) => `\x1b[1m${s}\x1b[0m`,
  dim: (s) => `\x1b[2m${s}\x1b[0m`,
  green: (s) => `\x1b[32m${s}\x1b[0m`,
};

console.log(c.bold("\nCPE 412 -- teaching plan, extracted from the DOCX\n"));
let n = 0;
for (const r of rows) {
  n++;
  const title = r.topics[0] ?? "(untitled)";
  console.log(c.green(`  [${String(n).padStart(2, "0")}] ${title}`) + c.dim(`   ${r.time}`));
  console.log(c.dim(`       CILO ${r.cilo || "-"}`));
  console.log(c.dim(`       ${r.outcomes.length} outcome(s), ${r.topics.length - 1} sub-topic(s)`));
}
console.log(
  c.dim(
    `\n  ${rows.length} plan rows, ` +
      `${rows.reduce((a, r) => a + r.outcomes.length, 0)} unit outcomes, ` +
      `${rows.reduce((a, r) => a + r.topics.length, 0)} topic lines\n`,
  ),
);
